src/base/file_utils.cc

```diff
@@ -400,12 +408,25 @@ base::Status ListFilesRecursive(const std::string& dir_path,
       }
       struct stat dirstat;
       std::string full_path = cur_dir + dirent->d_name;
-      PERFETTO_CHECK(stat(full_path.c_str(), &dirstat) == 0);
+
+      if (stat(full_path.c_str(), &dirstat) != 0) {
+        // skip broken symlinks.
+        if (errno == ENOENT) {
+          continue;
+        }
+        PERFETTO_FATAL("stat %s", full_path.c_str());
+      }
+
       if (S_ISDIR(dirstat.st_mode)) {
+        auto dir_id = std::make_pair(dirstat.st_dev, dirstat.st_ino);
+        if (visited_dirs.find(dir_id) != visited_dirs.end()) {
+          continue;
+        }
```

Comment:
This one makes me a little bit nervous. if for any reason (some weird filesystem, or selinux permissions) we read 0,0 here, we can end up ignoring legitimate dirs.

Could the check be done on the dir that we push ?